tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,144 @@
+            assert_script_run "virsh dumpxml --inactive $guest > $downloaded_xml_dir/$guest.xml";
+        }
+    }
+
+    foreach my $guest (keys %virt_autotest::common::guests) {
+
+        #irqbalance test run on sle12sp3+ guest
+        return if guest_is_sle($guest, '<=12-sp2');
+
+        record_info("Test $guest");
+        prepare_guest_for_irqbalance($guest);
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        #Check if network interrupts are distributed over CPUs in guest initally as irqbalance is enabled by default
+        my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            die("The value of network interrupts for CPU" . $cpu_id . " is 0") if ($initial_total_irqs_on_cpu[$cpu_id] == 0);
+        }
+
+        # Bind all irqs to cpu0 inside guest
+        # with irqbalance enabled, trying to bind IRQs to one CPU core would not succeed
+        my @nic_irqs_id = split('\n', script_output("ssh root\@$guest \"grep vif /proc/interrupts | cut -d ':' -f1 | sed 's/^ *//'\""));
+        my $cat_affinity_files_cmdline = "ssh root\@$guest \"cd /proc/irq && cat";
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+            $cat_affinity_files_cmdline .= " $irq_id/smp_affinity";
+        }
+        $cat_affinity_files_cmdline .= "\"";
+        my @affinity_after_binding = split('\n', script_output($cat_affinity_files_cmdline, 60));
+        foreach my $affinity (@affinity_after_binding) {
+            record_soft_failure("NIC IRQs were binded to one cpu core successfully, which is not expected.") if $affinity eq 1;
+        }
+
+        # Generate xen guest network load
+        #julie: need find a proper download
+        #julie: 4 sec to download this TW en.tlk in OSD
+        my $download_stuff = "http://download.opensuse.org/distribution/openSUSE-current/repo/oss/boot/x86_64/loader/en.tlk";
+        assert_script_run("ssh root\@$guest \"wget $download_stuff\"");
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
+        record_info("IRQs of downloads distribution", "@increased_irqs_on_cpu");
+    }
+
+}
+
+#set up guest test environment to run irqbalance test
+sub prepare_guest_for_irqbalance {
+    my $vm_name = shift;
+
+    #4 or more vcpu is needed
```

Comment:
why do not use with hotplug vpu here. 
For example:
`virsh setvcpus --domain $guest --count 4 --live`

Reply:
if it was not a guest with cpu=2/4, cpu hotplug would not work.